scripts/constituents.py

```diff
@@ -484,124 +468,168 @@ def write_host_routines(cap, host, reg_funcname, init_funcname, num_const_funcna
+            # end for
+            cap.write("num_consts = num_consts + num_dyn_consts", 2)
+            cap.comment(f"Pack {dyn_const_name} array", 2)
+            cap.write(f"allocate({dyn_const_name}(num_dyn_consts), stat={herrcode})", 2)
+            cap.write(f"if ({herrcode} /= 0) then", 2)
+            cap.write(f"{herrmsg} = 'failed to allocate {dyn_const_name}'", 3)
+            cap.write("return", 3)
+            cap.write("end if", 2)
+            cap.write("index_start = 0", 2)
+            for idx, scheme in enumerate(sorted(dyn_const_dict)):
+                cap.write(f"do index = 1, size(dyn_const_prop_{idx}, 1)", 2)
+                cap.write(f"{dyn_const_name}(index + index_start) = dyn_const_prop_{idx}(index)", 3)
+                cap.write("end do", 2)
+                cap.write(f"index_start = size(dyn_const_prop_{idx}, 1)", 2)
+                cap.write(f"deallocate(dyn_const_prop_{idx})", 2)
+            # end for
+        # end if
+        cap.comment("Initialize constituent data and field object", 2)
+        stmt = f"call {const_obj_name}%initialize_table(num_consts)"
+        cap.write(stmt, 2)
         # Register host model constituents
-        cap.comment("Add host model constituent metadata", 3)
-        cap.write("do index = 1, size(host_constituents, 1)", 3)
-        cap.write(f"if ({herrcode} == 0) then", 4)
-        cap.write("const_prop => host_constituents(index)", 5)
-        stmt = "call {}%new_field(const_prop, {})"
-        cap.write(stmt.format(const_obj_name, obj_err_callstr), 5)
-        cap.write("end if", 4)
-        cap.write("nullify(const_prop)", 4)
-        cap.write("if ({} /= 0) then".format(herrcode), 4)
-        cap.write("exit", 5)
-        cap.write("end if", 4)
-        cap.write("end do", 3)
-        cap.write("end if", 2)
-        cap.blank_line()
+        cap.comment("Add host model constituent metadata", 2)
+        cap.write("do index = 1, size(host_constituents, 1)", 2)
+        cap.write("const_prop => host_constituents(index)", 3)
+        stmt = f"call {const_obj_name}%new_field(const_prop, {obj_err_callstr})"
+        cap.write(stmt, 3)
+        cap.write("nullify(const_prop)", 3)
+        cap.write(f"if ({herrcode} /= 0) then", 3)
+        cap.write("return", 4)
+        cap.write("end if", 3)
+        cap.write("end do", 2)
+        cap.blank_line()
+        # Register dynamic constituents
+        if len(dyn_const_dict) > 0:
+           cap.comment("Add dynamic constituent properties", 2)
+           cap.write(f"do index = 1, size({dyn_const_name}, 1)", 2)
+           cap.write(f"const_prop => {dyn_const_name}(index)", 3)
+           stmt = f"call {const_obj_name}%new_field(const_prop, {obj_err_callstr})"
+           cap.write(stmt, 3)
+           cap.write("nullify(const_prop)", 3)
+           cap.write(f"if ({herrcode} /= 0) then", 3)
+           cap.write("return", 4)
+           cap.write("end if", 3)
+           cap.write("end do", 2)
+        # end if
+        
```

Comment:
we can't deallocate {dyn_const_name} here because that's the module-level variable that holds the data for the dynamic constituents; deallocating it would mean you could no longer access those constituent properties